Add docker configuration for quick local setup

pacodelaluna opened the pull request:
#### What? Why?
Let's add a Docker setup for local environment. The idea behind it is to allow developers or testers to setup quickly there environment.

Closes #2290. Already closed by added for reference.


It will add great value to the project I think, people will be able to quickly see how the app is working and then focus their energy on real issues. On the downside, we will have to maintain this configuration but we have also to consider it as documentation, and almost real-time-updated one, so I think it worth it.


#### What should we test?
I have added a `DOCKER.md` file which describes the process to follow in order to have the application working locally. Tell me if it needs adjustments.



#### Release notes
- Add Docker configuration for quick local setup



#### Changelog Category: Added



#### Documentation updates
I will add a link to the `DOCKER.md` file in the wiki once the PR is validated and merged.

pacodelaluna commented on Dockerfile:
```diff
@@ -0,0 +1,57 @@
+RUN apt-get update
+
+## Rbenv & Ruby part
+
```
I could't use the `ruby-2.1.5` Docker image as base because it is based on an old Ubuntu which is not maintained anymore.

sauloperez replied:
I did a quick try the other day coming from the gathering and I think I did manage to get that image running. However, what you mention of being out-of-date is a valid point. However, I'd like to try starting from an Alpine Linux because it'll lead to a much more lightweight image.

pacodelaluna replied:
Let's try with Alpine then, I was trying to use the images the closest to our prod, but lightness is a big concern too.

Matt-Yorkley replied:
> I have tried to use alpine but it is giving me issues with PhantomJS and also rbenv compilation.

I don't think it's worth spending more time getting Alpine to work. It's only for elite Docker-Hipsters anyway. :wink:

pacodelaluna commented:
@sauloperez @luisramos0 @Matt-Yorkley @mkllnk It is still not a onliner setup, but I think it can help devs for onboarding. It will possibly require some adjustments so don't hesitate to tell me.

sauloperez commented on DOCKER.md:
```diff
@@ -0,0 +1,66 @@
+
+Clone the repository:
+
+    git clone git@github.com:openfoodfoundation/openfoodnetwork.git
```
I'd rather use backticks (```) to make use of markdown's code blocks and precede commands with `$` to make it more explicit. They are conventions I always follow in the wiki.

pacodelaluna replied:
Ok, I will check this out, for now I just followed how `GETTING_STARTED.md` is written.

sauloperez replied:
Good to know. Might be worth updating it

sauloperez commented on lib/tasks/dev.rake:
```diff
@@ -1,5 +1,26 @@
+
+      p '== Migrate'
+      Rake::Task['db:migrate'].invoke
+    end
```
I don't see the value of abstracting this. Can't we simply run `bundle exec db:setup` in `docker-compose.yml`? It does the `db:create` and `db:seed` for us. If we prefer, we could simply run `bundle exec db:create db:schema:load` instead.

Besides, there is no way around learning what each of these Rails tasks do is someone wants to develop, IMO. Hiding them does no good to new contributors.

pacodelaluna replied:
The value is that we may not want to rebuild the dataset each time we are launching the container, especially that there are few tests that can take time to setup at data level, and you may need to restart the container for some reasons (Data migration, dependencies loading, ...). Also, it is not straight forward to load the data today and you unfortunately need to load them manually. 

This task will provide some flexibility to the user who wants to perform some tasks launching the container, that's why I let some commented lines here too. It is not perfect, but it is a first step, I am always putting myself in the shoes of someone launching the app for the first time, and how to make it smoothly.

This was just to clarify my point of this weird task here, but if it is a real concern for you, I will adopt the strategy you propose, no problem.

sauloperez replied:
let's see what others think. I would prefer teaching new devs to fish rather than giving ourselves the fish.

Matt-Yorkley replied:
Yeah, I'd remove this. The rake tasks already exist, right? This logic can just go in the docker files. See my comment above for running `db:seed` non-interactively.

sauloperez commented on docker-compose.yml:
```diff
@@ -0,0 +1,35 @@
+
+services:
+  db:
+    image: postgres:9.3
```
I believe we run 9.5 in most instances, although we don't lock it :see_no_evil:

sauloperez commented on config/database.docker.yml:
```diff
@@ -0,0 +1,20 @@
+  <<: *default
+  database: open_food_network_test
+  username: ofn
+  password: f00d
```
I'd rather work on https://github.com/openfoodfoundation/ofn-install/issues/387 (which I really want to do soon) which will provide the needed flexibility with ENV vars so we don't need to add purpose-specific database.yml like this one.

pacodelaluna replied:
Ok, let's wait for it then.

Matt-Yorkley replied:
#387 would be nice, but I don't think it needs to block this Docker setup..?

sauloperez replied:
We can go with https://github.com/openfoodfoundation/openfoodnetwork/pull/3954

sauloperez commented on Dockerfile:
```diff
@@ -0,0 +1,57 @@
+RUN curl -sL https://deb.nodesource.com/setup_8.x | bash - && \
+    curl -sS https://dl.yarnpkg.com/debian/pubkey.gpg | apt-key add - && \
+    echo "deb https://dl.yarnpkg.com/debian/ stable main" | tee /etc/apt/sources.list.d/yarn.list
+RUN apt-get update && apt-get install -y nodejs yarn
```
I'd great to run them with a single `RUN` command to creating multiple intermediate images and the disk space that takes.

pacodelaluna replied:
I couldn't see these intermediate images while building the app, but I could see that `RUN` is helping stepping the image construction. I gave still reduce the number of calls to 3.

sauloperez commented on Dockerfile:
```diff
@@ -0,0 +1,57 @@
+RUN apt-get install -yqq --no-install-recommends build-essential git-core curl zlib1g-dev libssl-dev libreadline-dev libyaml-dev libxml2-dev libxslt1-dev libcurl4-openssl-dev libffi-dev phantomjs
+
+# System dependencies
+RUN apt-get install -y wait-for-it
```
See my comment about `RUN` above. I believe each of these commands creates a separate image that Docker caches.

sauloperez commented on DOCKER.md:
```diff
@@ -0,0 +1,66 @@
+### Docker
+
+It is possible to setup the OpenFoodnetwork app easily with Docker and Docker Compose.
```
OpenFoodnetwork => OpenFoodNetwork or Open Food Network

sauloperez commented on DOCKER.md:
```diff
@@ -0,0 +1,66 @@
+
+Please check the documentation here, https://docs.docker.com/install/ to install Docker.
+
+For Docker Compose, information are here: https://docs.docker.com/compose/install/.
```
information is here :ok_hand:

pacodelaluna commented:
I have fixed many remarks, but there are still 3 points to consider:

1. Use of `alpine` base instead `ubuntu` one: I have tried to use alpine but it is giving me issues with PhantomJS and also rbenv compilation. I don't want to spend more time on it as it will just save some MBs on hard drive, I feel like spending this time coding but if someone want to take a look with me, we can do it together.

2. Avoid specific rake task for setup: I can't figure out a fast setup without this task as imho `docker-compose` should run the migrations, and migrations are not running properly from a fresh db. It is not ideal but it is working. And this setup is not only for devs, it should help anybody to setup the env locally.

3. Remove the Docker-specific database.yml file: This task is linked to an other one, so we can wait for it, or update the Docker configuration once the other task is done (openfoodfoundation/ofn-install#387), it depends how long it will take.

@sauloperez Please tell your opinion about these points.

mkllnk commented on DOCKER.md:
```diff
@@ -0,0 +1,82 @@
+
+### Notes
+
+- It was not possible to integrate the seeding part directly because we need the input of the user and Docker Compose was not allowing it. Need to fix this.
```
Can you set environment variables? Setting `AUTO_ACCEPT` should skip the username and password question.

Matt-Yorkley replied:
:100: for `ENV['AUTO_ACCEPT'] = true`

Then we can run the db:seed task without any prompts, and the default login details will be: `spree@example.com` and `spree123`. Alternatively you can set `ENV['ADMIN_EMAIL']` and `ENV['ADMIN_EMAIL']`.

So that means both the `db:seed` and `ofn:sample_data` tasks can both be run automatically during the setup, and there's no need for manual steps. :+1:

Matt-Yorkley replied:
> It is still not a 'oneliner' setup

We can do it! :muscle: :smile:

pacodelaluna replied:
@Matt-Yorkley Thanks for the trick! I will try it asap.

@mkllnk I agree that we could put this outside this repository, just saw the Vagrant image made by @Matt-Yorkley few days ago, we could do something similar.

luisramos0 replied:
is the seeding part integrated now? we need to remove this line if that is the case.

mkllnk commented on Dockerfile:
```diff
@@ -0,0 +1,32 @@
+ENV BUNDLE_PATH /bundles
+
+COPY . /usr/src/app/
+COPY Gemfile* /usr/src/app/
```
Why do we need to copy the Gemfile separately?

pacodelaluna replied:
Good question :) I will check this out.

mkllnk commented on config/application.docker.yml:
```diff
@@ -0,0 +1,64 @@
+# Add application configuration variables here, as shown below.
```
Why not use the example file?

pacodelaluna replied:
Good idea! It will reduce `docker` footprint.

Matt-Yorkley replied:
Are there any variables that need to be customised?

Matt-Yorkley commented on Dockerfile:
```diff
@@ -0,0 +1,32 @@
+    gem install bundler --version=1.17.2
+
+# Postgres
+RUN sh -c "echo 'deb http://apt.postgresql.org/pub/repos/apt/ precise-pgdg main' > /etc/apt/sources.list.d/pgdg.list" && \
```
Shouldn't this be `bionic-pgdg main` instead of `precise-pgdg main` if you're using the `ubuntu:18.04` image?

sauloperez commented:
I still want to give a try and share feedback. I might switch from my current lxc setup.

Matt-Yorkley commented on Dockerfile:
```diff
@@ -0,0 +1,32 @@
+FROM ubuntu:18.04
+
+#Install all the requirements
+RUN apt-get update && apt-get install -y curl git build-essential software-properties-common wget zlib1g-dev libssl1.0-dev libreadline-dev libyaml-dev libffi-dev libxml2-dev libxslt1-dev phantomjs wait-for-it
```
I think we don't use `phantomjs` anymore.

pacodelaluna replied:
I will try to remove it then, was getting it from the wiki.

Matt-Yorkley commented:
I'm getting build errors at the `rbenv` step, and I think they look familiar... https://github.com/openfoodfoundation/ofn-install/pull/348/commits/bbd03fe7a7f66b0c147aab8e0a3028db67a95c30#diff-3cd6f6e3b6e4c2ea1695765d4b707c63R47

Matt-Yorkley commented:
I just had a play with this, it's pretty awesome. I made some tweaks here, the `command` section is a bit hacky but it works well: https://github.com/pacodelaluna/openfoodnetwork/compare/add-docker-configuration-for-dev-setup...Matt-Yorkley:vamos

No manual setup steps :muscle:

sauloperez commented:
> So does this mean we can create a downloadable Docker image from this, and devs won't have to build Ruby with rbenv again? 

:heart:

sauloperez commented:
LOL the branch name @Matt-Yorkley :rofl: I do prefer it the way you suggest.

sauloperez commented on config/database.docker.yml:
```diff
@@ -0,0 +1,19 @@
+  adapter: postgresql
+  encoding: utf8
+  pool: 5
+  timeout: 5000
```
why do we need that?

sauloperez commented on config/database.docker.yml:
```diff
@@ -0,0 +1,19 @@
+  pool: 5
+  timeout: 5000
+  host: db
+  port: 5432
```
we surely don't need that

pacodelaluna commented:
@Matt-Yorkley Great, you fixed almost all the comments. I add you on my repository as collaborator, or you can create a new PR. Let's work on the same branch to avoid merge issues. I will work on the @mkllnk remarks about the use of example files in Docker Compose.
I think that with this, it will just be 3 Docker files and 1 MD file, which is a pretty reduce footprint, don't know if we should still put it outside the repository. @Matt-Yorkley @sauloperez @luisramos0 What do you think about it?

luisramos0 commented:
we will fine with this in ofn repo or outside, I'd vote to keep it inside the repo if only a couple of new files.
Docker is quite fancy, and not just that, it's a good sign we use docker and I think that keeping it in ofn repo makes it more visible and will probably make more people use it and improve it.

mkllnk commented:
It can stay in the repository if it's only a couple of files. Or we create a docker folder for the files in the repository. I'm just aware that we have many files for many different tools and it can get a bit messy.
```
# current master
$ ls -a
app
bin
.bundle
.codeclimate.yml
config
config.ru
CONTRIBUTING.md
coverage
db
engines
Gemfile
Gemfile.lock
GETTING_STARTED.md
.git
.github
.gitignore
Guardfile
knapsack_rspec_report.json
lib
LICENCE.txt
log
.mailmap
nbproject
node_modules
.node-version
package.json
public
Rakefile
README.md
reports
.rspec
.rubocop_manual_todo.yml
.rubocop_styleguide.yml
.rubocop_todo.yml
.rubocop.yml
.ruby-gemset
.ruby-version
script
.scss-lint.yml
spec
tmp
.travis.yml
.tx
vendor
zeus.json
```

amazzoccone commented:
+1 @pacodelaluna ! 🎉  There are multiple advantages from implement **Docker**, both in local and production envs.
- Local: quick setup and standard env (everyone has a more similar environment, except for os of course) are the first ones, helping new contributors as well and no need to force developer to install the stack (db, cache, language, etc).
- Stage/Production: useful in the **integration** (preparation of the deployment of a new release), simplifying a lot of the process with the usage of images and a registry, etc etc.

In my opinion, at least at the beginning would be great to have the choice of Docker in local env.

luisramos0 commented:
welcome to OFN @amazzoccone 
join us on slack 🎸 https://openfoodnetwork.org/slack-invite

Matt-Yorkley commented:
@pacodelaluna I submitted my suggestions as a PR on your branch :+1:

luisramos0 commented on DOCKER.md:
````diff
@@ -0,0 +1,82 @@
+$ bundle exec rake db:seed
+```
+
+Load thre sample data from Open Food Network:
````
Load THE sample

luisramos0 commented on Dockerfile:
```diff
@@ -0,0 +1,32 @@
+FROM ubuntu:18.04
+
+#Install all the requirements
```
space after #

luisramos0 commented on Dockerfile:
```diff
@@ -0,0 +1,32 @@
+#Install all the requirements
+RUN apt-get update && apt-get install -y curl git build-essential software-properties-common wget zlib1g-dev libssl1.0-dev libreadline-dev libyaml-dev libffi-dev libxml2-dev libxslt1-dev wait-for-it
+
+#Setup ENV variables
```
space after #

you can probably run rubocop on this file as well to get these things flagged.

luisramos0 commented:
I forgot the most important part of my review: this is amazing stuff @pacodelaluna really 👏

mkllnk commented:
@pacodelaluna Do you have time to fix those typos or would you like someone else to take over? We can also start a tech-debt issue for the remaining questions (remove unneeded files). You've done a great job already.

pacodelaluna commented:
Sorry guys, very busy nowadays but I will do all I can this weekend and keep you updated.

pacodelaluna commented:
I have rebased and fixed all the comments, but I had a last issue with the Sample Data setup. It was failing while relaunching the container, some constraints on `Customer` object on `db:setup`. I have decided to reset the db instead but then we are having an issue with `ofn:sample_data`, `Spree::User` object is not found inside the environment (possibly because `db:reset` is flushing it).
I fixed this with the next commit, but please tell me if there is an issue. Not sure it is so good to reset data all the time, but we can tweak this easily.
Anyway, I guess we will have to work on Sample Data soon or later, but it will be an other topic.

mkllnk commented on Dockerfile:
```diff
@@ -0,0 +1,31 @@
+
+# Postgres
+RUN sh -c "echo 'deb http://apt.postgresql.org/pub/repos/apt/ bionic-pgdg main' > /etc/apt/sources.list.d/pgdg.list" && \
+    wget --quiet -O - http://apt.postgresql.org/pub/repos/apt/ACCC4CF8.asc | apt-key add - && \
```
I would definitely use the https URL to obtain a key. Otherwise the whole cryptography with signatures is useless.

pacodelaluna replied:
Totally right, just updated it.

mkllnk commented on config/application.example.yml:
```diff
@@ -0,0 +1,64 @@
+# Add application configuration variables here, as shown below.
```
This is exactly the same file as `config/application.yml.example`. Can't we use the existing file? Or rename it to this one if we need to? That would just need some other changes in other places.

pacodelaluna replied:
Right, forgot to delete the old one, just updated it.

luisramos0 commented:
are you renaming the sample file on purpose or is it just a mistake?
config/application.yml.example → config/application.example.yml 

if you want/need to rename it, you will need to update a few more things, see the broken build.

pacodelaluna commented:
@luisramos0 Yes, I am renaming it as it will be clearer for syntax detection, what to you think about it? Just fixed the setup script, I guess it was the spec issue.

luisramos0 commented:
I see why you want to rename the file but you may be creating trouble with this.
you need to change the semaphore build config here: https://semaphoreci.com/openfoodfoundation/openfoodnetwork-2/settings
and also you need to update ofn-install/scripts/db-setup.sh as well

maybe easier to leave it as it was?

pacodelaluna commented:
Sure, touched a Pandora box here :) I put back as it was.

luisramos0 commented:
sorry, still not correct, you are deleting config/application.yml.example now

pacodelaluna commented:
@luisramos0 Sorry, forgot to add the file, fixed now. I think we are good now.

mkllnk commented on docker-compose.yml:
```diff
@@ -0,0 +1,38 @@
+      - .:/usr/src/app
+      - gems:/bundles
+      - ./config/database.yml:/usr/src/app/config/database.yml
+      - ./config/application.example.yml:/usr/src/app/config/application.yml
```
Don't you need to change this line now to use the old file `application.yml.example`?

pacodelaluna replied:
Good catch, I have rebuild the image, the container is running, hope it is good now...

luisramos0 commented:
@pacodelaluna 👏 👏 👏
